x/wasm/keeper/gas_register.go

```diff
@@ -0,0 +1,117 @@
+//
+// Please note that all gas prices returned to the wasmer engine should have this multiplied
+const DefaultGasMultiplier uint64 = 100
+
+// DefaultInstanceCost is how much SDK gas we charge each time we load a WASM instance.
+// Creating a new instance is costly, and this helps put a recursion limit to contracts calling contracts.
+const DefaultInstanceCost uint64 = 40_000
+
+// DefaultCompileCost is how much SDK gas we charge *per byte* for compiling WASM code.
+const DefaultCompileCost uint64 = 2
+
+// DefaultEventAttributeCost is how much SDK gas we charge *per byte* for attribute data in events
+const DefaultEventAttributeCost uint64 = 10
+
+type GasRegister struct {
+	instanceCost  sdk.Gas
+	compileCost   sdk.Gas
+	gasMultiplier sdk.Gas
+
+	eventAttributeCountCost  sdk.Gas
+	eventAttributeLengthCost sdk.Gas
+	freeTierAttributeLength  int
+}
+
+func DefaultGasRegister() GasRegister {
+	return GasRegister{
+		instanceCost:             DefaultInstanceCost,
+		compileCost:              DefaultCompileCost,
+		gasMultiplier:            DefaultGasMultiplier,
+		eventAttributeCountCost:  0,
+		eventAttributeLengthCost: DefaultEventAttributeCost,
+		freeTierAttributeLength:  0,
+	}
+}
+func NewGasRegister(instanceCost sdk.Gas, compileCost sdk.Gas, gasMultiplier sdk.Gas, eventAttributeCountCost sdk.Gas, eventAttributeLengthCost sdk.Gas, freeTierAttributeLength int) GasRegister {
+	return GasRegister{
+		instanceCost:             instanceCost,
+		compileCost:              compileCost,
+		gasMultiplier:            gasMultiplier,
+		eventAttributeCountCost:  eventAttributeCountCost,
+		eventAttributeLengthCost: eventAttributeLengthCost,
+		freeTierAttributeLength:  freeTierAttributeLength,
+	}
+}
+
+func (g GasRegister) NewContractInstanceCost(pinned bool, msgLen int, labelLength int) storetypes.Gas {
+	return g.InstantiateContractCost(pinned, msgLen)
+}
+
+func (g GasRegister) CompileCost(byteLength int, sourceCodeUrlLen int, builderLen int) storetypes.Gas {
+	return g.compileCost * uint64(byteLength)
+}
+
+func (g GasRegister) InstantiateContractCost(pinned bool, msgLen int) sdk.Gas {
+	if pinned {
+		return 0
+	}
+	return g.instanceCost
+}
+
+func (g GasRegister) ReplyCost(pinned bool, reply wasmvmtypes.Reply) sdk.Gas {
+	var eventGas sdk.Gas
+	msgLen := len(reply.Result.Err)
+	if reply.Result.Ok != nil {
+		msgLen += len(reply.Result.Ok.Data)
+		for _, e := range reply.Result.Ok.Events {
+			msgLen += len(e.Type)
+			eventGas += g.EventCosts(e.Attributes)
+		}
+	}
+	return eventGas + g.InstantiateContractCost(pinned, msgLen)
+}
+
+func (g GasRegister) EventCosts(evts []wasmvmtypes.EventAttribute) sdk.Gas {
+	if len(evts) == 0 {
+		return 0
+	}
+	var storedBytes int
+	for _, l := range evts {
+		storedBytes += len(l.Key) + len(l.Value)
+	}
+	// apply free tier
+	if storedBytes <= g.freeTierAttributeLength {
+		storedBytes = 0
+	} else {
+		storedBytes -= g.freeTierAttributeLength
+	}
+	// total Length * costs + attribute count * costs
```

Comment:
very nice formula.

Reply:
Using `sdk.Int` to catch overflow?

Reply:
ja, not sure if this required... but better safe than sorry